File: Zig/day01/src/main.zig
const PUZZLE: []const u8 = @embedFile("../../../Inputs/day01.txt");

usingnamespace @import("std");
usingnamespace @import("fuel.zig");

fn part1(s: []const u8) i64 {
    var lines = mem.separate(s, "\r\n");

    var result = i64(0);
    while (lines.next()) |line| {
        var mass = fmt.parseInt(i64, line, 10) catch @panic("Failed to parse int");

        result += @divFloor(mass, 3) - 2;
    }

    return result;
}

fn part2(s: []const u8) i64 {
    var lines = mem.separate(s, "\r\n");

    var result = i64(0);

    while (lines.next()) |line| {
        var mass = fmt.parseInt(i64, line, 10) catch @panic("Failed to parse int");

        var fueliter = FuelIterator.new(mass);

        while (fueliter.next()) |fuel| {
            result += fuel;
        }
    }

    return result;
}

pub fn main() anyerror!void {
    @setEvalBranchQuota(15000);
    comptime var p1 = part1(PUZZLE);
    comptime var p2 = part2(PUZZLE);

    debug.warn("Part 1: {}\nPart 2: {}", p1, p2);
}

File: Rust/day01/src/main.rs
static PUZZLE: &'static str = include_str!(r"..\..\..\Inputs\day01.txt");

mod fuel;
use fuel::Fuel;

use std::iter;

fn parse<'a>(input: &'a str) -> impl Iterator<Item = i64> + 'a {
    input.lines().map(|line| line.parse::<i64>().unwrap())
}

fn part1(s: &str) -> i64 {
    parse(s).map(|mut mass| mass.fuel().unwrap()).sum::<i64>()
}

fn part2(s: &str) -> i64 {
    parse(s)
        .flat_map(|mut mass| iter::from_fn(move || mass.fuel()))
        .sum()
}

fn main() {
    let p1 = part1(PUZZLE);
    let p2 = part2(PUZZLE);

    println!("Part 1: {}\nPart 2: {}", p1, p2);
}
